src/Common/Internal/Authentication/SharedAccessSignatureHelper.php

```diff
@@ -0,0 +1,277 @@
+
+        // check that expiracy is valid
+        Validate::isString($signedExpiracy, 'signedExpiracy');
+        Validate::notNullOrEmpty($signedExpiracy, 'signedExpiracy');
+        Validate::isDateString($signedExpiracy, 'signedExpiracy');
+
+        // check that signed start is valid
+        Validate::isString($signedStart, 'signedStart');
+        if(strlen($signedStart) > 0)
+        {
+            Validate::isDateString($signedStart, 'signedStart');
+        }
+
+        // check that signed IP is valid
+        Validate::isString($signedIP, 'signedIP');
+
+        // validate and sanitize signed protocol
+        $signedProtocol = $this->validateAndSanitizeSignedProtocol($signedProtocol);
+
+        // construct an array with the parameters to generate the shared access signature at the account level
+        $parameters = array();
+        $parameters[] = $this->accountName;
+        $parameters[] = urldecode($signedPermissions);
+        $parameters[] = urldecode($signedService);
+        $parameters[] = urldecode($signedResourceType);
+        $parameters[] = urldecode($signedStart);
+        $parameters[] = urldecode($signedExpiracy);
+        $parameters[] = urldecode($signedIP);
+        $parameters[] = urldecode($signedProtocol);
+        $parameters[] = urldecode($signedVersion);
+
+        // implode the parameters into a string
+        $stringToSign = utf8_encode(implode("\n", $parameters)."\n");
+
+        // decode the account key from base64
+        $decodedAccountKey = base64_decode($this->accountKey);
+        
+        // create the signature with hmac sha256
+        $signature = hash_hmac("sha256", $stringToSign, $decodedAccountKey, true);
+
+        // encode the signature as base64
+        $base64Signature = base64_encode($signature);
+
+        // return the signature
+        return $base64Signature;
+    }
+
+    function validateAndSanitizeSignedService($signedService) {
+        // validate signed service is not null or empty
+        Validate::isString($signedService, 'signedService');
+        Validate::notNullOrEmpty($signedService, 'signedService');
+
+        // sanitize signed service
+        $sanitizedSignedService = $this->removeDuplicateCharacters(strtolower($signedService));
+        
+        // The signed service should only be a combination of the letters b(lob) q(ueue) t(able) or f(ile)
+        $signedServiceIsValid = preg_match("/^[bqtf]*$/", $sanitizedSignedService);
+        if(!$signedServiceIsValid) {
+            throw new \InvalidArgumentException(Resources::SIGNED_SERVICE_INVALID_VALIDATION_MSG);
+        }
+
+        return $sanitizedSignedService;
+    }
+
+    function validateAndSanitizeSignedResourceType($signedResourceType) {
+        // validate signed resource type is not null or empty
+        Validate::isString($signedResourceType, 'signedResourceType');
```

Comment:
Yes, I agree but this private methods are validators for the public one. It validates the input parameters and business rules for generateAccountSharedAccessSignature.